### PR TITLE
Catch and fix negative volumes of triangles/quads in 2D Gmsh files.

### PR DESCRIPTION
**Pull request by jmark**
**_Describe your changes here:_**

This PR checks for negative volumes of triangles and quads and corrects this when reading Gmsh files. "Positive" 2D shapes are defined by the cross product of the two sides of the triangle/quad pointing towards the same direction as the positive z-direction. Of course, this only works when the quad/triangle lies in the x-y plane.

You can use this Gmsh files with negative quads for testing: https://gist.githubusercontent.com/jmark/bfe0d45f8e369298d6cc637733819013/raw/cecf86edecc736e8b3e06e354c494b2052d41f7a/rectangle_with_negative_volumes.msh

~~This PR has one problem though: It breaks the Lagrange-Geometry handler for Hexes since some faces of a hex (which are quads) are created with negative volumes. I hope this can be easily fixed. This has probably something to do with the fact that the faces resp. quads do not lie in the x-y-plane anymore.~~

**_All these boxes must be checked by the reviewers before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is fully understood, bug free, well-documented and well-structured.

#### General
- [ ] The reviewer executed the new code features at least once and checked the results manually

- [ ] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [ ] New source/header files are properly added to the Makefiles
- [ ] The code is well documented
- [ ] All function declarations, structs/classes and their members have a proper doxygen documentation
- [ ] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)

#### Tests
- [ ] The code is covered in an existing or new test case using Google Test

#### Github action

- [ ] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)
- [ ] All tests pass (in various configurations, this should be executed automatically in a github action)

  If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually)

#### Scripts and Wiki

- [ ] If a new directory with source-files is added, it must be covered by the `script/find_all_source_files.scp` to check the indentation of these files.
- [ ] If this PR introduces a new feature, it must be covered in an example/tutorial and a Wiki article.

#### Licence

- [ ] The author added a BSD statement to `doc/` (or already has one)

### CODE REVIEW

**Comment by jmark** [comment]
@sandro-elsweijer @CsatiZoltan Would be please take a look at this PR. Is there a fix for the Lagrange geometry handler?

**Comment by jmark** [comment]
> @sandro-elsweijer @CsatiZoltan Would be please take a look at this PR? Is there a fix for the Lagrange geometry handler?

Nevermind. I just got an idea. Let me try that out first.

**Comment by CsatiZoltan** [comment]
@jmark The Lagrange geometry elements were designed to have positive face orientations, i.e. with outward unit vector.

**Comment by jmark** [comment]
> @jmark The Lagrange geometry elements were designed to have positive face orientations, i.e. with outward unit vector.
> 
> > This has probably something to do with the fact that the faces resp. quads do not lie in the x-y-plane anymore.
> 
> I think so.
> 
> > This PR checks for negative volumes of triangles and quads. That is, the cross product of the two sides of the triangle/quad must point towards the same direction as the positive z-direction.
> 
> What you test for is negative _area_, no? If your goal is to test positive orientation for tri/quad, then you want to work in 2D. In that case, in my opinion, you do not have to test the faces of 3D elements. Those should rather be subject to a positivity test for volumes (in which case, you could use the triple product of three element edges). Do I miss something?

@CsatiZoltan You got it right. The negative volume test for 2D shapes makes only really sense in the x-y plane. Anyway,
I removed the check in the validation routines. So no code breaking anymore.

Instead I only check for "negative" 2D elements when reading 2D Gmsh files. This is the new purpose of this PR.

File: src/t8_cmesh/t8_cmesh.cxx [hunk 1]
```diff
@@ -479,22 +479,35 @@ t8_cmesh_set_tree_class (t8_cmesh_t cmesh, t8_gloidx_t gtree_id, t8_eclass_t tre
...
+   *      z             For 2D meshes we enforce the right-hand-rule in terms
+   *      |             of node ordering. The volume is defined by the parallelepiped
+   *      | 2- - -(3)   spanned by the vectors between nodes 0:1 and 0:2 as well as the
+   *      |/____ /      unit vector in z-direction. This definition works for both triangles and quads.
+   *      0     1
+   *
+   *      6 ______  7   For Hexes and pyramids, if the vertex 4 is below the 0-1-2-3 plane,
    *       /|     /     the volume is negative. This is the case if and only if
    *    4 /_____5/|     the scalar product of v_4 with the cross product of v_1 and v_2 is
    *      | | _ |_|     smaller 0:
```
Comment (Prasanna-Ponnusamy) [inline src/t8_cmesh/t8_cmesh.cxx:513]: smaller than 0

**Review by Prasanna-Ponnusamy — APPROVED** [review]

**Pull request merged**